src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp

```diff
@@ -0,0 +1,158 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "ApparentHorizons/ChangeCenterOfStrahlkorper.hpp"
+
+#include <cmath>
+
+#include "ApparentHorizons/Strahlkorper.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "NumericalAlgorithms/RootFinding/TOMS748.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+namespace change_expansion_center_of_strahlkorper_detail {
+
+// Get the r_hat vector of the Strahlkorper, which defines the
+// collocation points.
+template <typename Frame>
+tnsr::i<DataVector, 3, Frame> get_rhat(
+    const Strahlkorper<Frame>& strahlkorper) noexcept {
+  const auto new_theta_phi = strahlkorper.ylm_spherepack().theta_phi_points();
+  auto r_hat_l =
+      make_with_value<tnsr::i<DataVector, 3, Frame>>(new_theta_phi[0], 0.0);
+  const DataVector sin_theta = sin(new_theta_phi[0]);
+  get<0>(r_hat_l) = sin_theta * cos(new_theta_phi[1]);
+  get<1>(r_hat_l) = sin_theta * sin(new_theta_phi[1]);
+  get<2>(r_hat_l) = cos(new_theta_phi[0]);
+  return r_hat_l;
+}
+
+// Work function called by the two interface functions.
+template <typename Frame>
+void change_expansion_center(
+    const gsl::not_null<Strahlkorper<Frame>*> strahlkorper,
+    const std::array<double, 3>& new_center,
+    const tnsr::i<DataVector, 3, Frame>& r_hat) noexcept {
+
+  // Get new_center minus old_center.
+  const auto& old_center = strahlkorper->center();
+  const std::array<double, 3> center_difference{
+      {new_center[0] - old_center[0], new_center[1] - old_center[1],
+       new_center[2] - old_center[2]}};
+
+  // For bracketing the root, get the min and max radius with respect
+  // to the new center.
+  const auto [r_min, r_max] =
+      [&center_difference, &r_hat, &strahlkorper ]() noexcept {
+    const auto radius_old = strahlkorper->ylm_spherepack().spec_to_phys(
```

Comment:
Can this be a const reference to avoid the allocation?

Reply:
I think you get an allocation no matter what, same as for theta_phi_points() above.